smithy-aws-endpoints/src/main/java/software/amazon/smithy/rulesengine/aws/traits/StandardPartitionalEndpointsTrait.java

```diff
@@ -0,0 +1,177 @@
+        super(ID, builder.getSourceLocation());
+        partitionEndpointSpecialCases = builder.partitionEndpointSpecialCases.copy();
+        endpointPatternType = Objects.requireNonNull(builder.endpointPatternType);
+    }
+
+    /**
+     * Gets the map of partition string to a list of partition endpoint special cases defined in the trait.
+     *
+     * @return Returns a map of partition string to a list of {@link PartitionEndpointSpecialCase}
+     */
+    public Map<String, List<PartitionEndpointSpecialCase>> getPartitionEndpointSpecialCases() {
+        return partitionEndpointSpecialCases;
+    }
+
+    /**
+     * Gets the endpoint pattern type defined in the trait.
+     *
+     * @return Returns a {@link EndpointPatternType}
+     */
+    public EndpointPatternType getEndpointPatternType() {
+        return endpointPatternType;
+    }
+
+    @Override
+    protected Node createNode() {
+        ObjectNode.Builder partitionEndpointSpecialCasesNodeBuilder = ObjectNode.objectNodeBuilder();
+        for (Map.Entry<String, List<PartitionEndpointSpecialCase>> entry
+                : partitionEndpointSpecialCases.entrySet()) {
+            List<Node> nodes = new ArrayList<>();
+            for (PartitionEndpointSpecialCase partitionEndpointSpecialCase : entry.getValue()) {
+                nodes.add(partitionEndpointSpecialCase.toNode());
+            }
+            partitionEndpointSpecialCasesNodeBuilder.withMember(entry.getKey(), Node.fromNodes(nodes));
+        }
+
+        return ObjectNode.objectNodeBuilder()
+                .sourceLocation(getSourceLocation())
+                .withMember(PARTITION_ENDPOINT_SPECIAL_CASES, partitionEndpointSpecialCasesNodeBuilder.build())
+                .withMember(ENDPOINT_PATTERN_TYPE, endpointPatternType.getName())
+                .build();
+    }
+
+    @Override
+    public SmithyBuilder<StandardPartitionalEndpointsTrait> toBuilder() {
+        return new Builder()
+                .partitionEndpointSpecialCases(partitionEndpointSpecialCases)
+                .endpointPatternType(endpointPatternType);
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    public static final class Provider extends AbstractTrait.Provider {
+        public Provider() {
+            super(ID);
+        }
+
+        @Override
+        public Trait createTrait(ShapeId target, Node value) {
+            ObjectNode objectNode = value.expectObjectNode();
+
+            EndpointPatternType endpointPatternType = EndpointPatternType
+                .fromNode(objectNode.expectStringMember(ENDPOINT_PATTERN_TYPE));
+
+            StandardPartitionalEndpointsTrait.Builder builder = builder()
+                    .sourceLocation(value)
+                    .endpointPatternType(endpointPatternType);
+
+            Optional<ObjectNode> memberNode = objectNode.getObjectMember(PARTITION_ENDPOINT_SPECIAL_CASES);
```

Comment:
Should use `objectNode.containsMember` in the if statement, since this `Optional` value is discarded.